File: core/src/main/java/com/alonsoruibal/chess/evaluation/Evaluator.java
package com.alonsoruibal.chess.evaluation;

import com.alonsoruibal.chess.Board;
import com.alonsoruibal.chess.Color;
import com.alonsoruibal.chess.bitboard.AttacksInfo;
import com.alonsoruibal.chess.bitboard.BitboardAttacks;
import com.alonsoruibal.chess.util.StringUtils;

public abstract class Evaluator {
	public static final int W = Color.W;
	public static final int B = Color.B;

	public static final int NO_VALUE = Short.MAX_VALUE;
	public static final int MATE = 30000;
	public static final int KNOWN_WIN = 20000;
	public static final int DRAW = 0;

	public static final int PAWN = 100;
	public static final int KNIGHT = 325;
	public static final int BISHOP = 325;
	public static final int ROOK = 500;
	public static final int QUEEN = 975;
	public static final int[] PIECE_VALUES = {0, PAWN, KNIGHT, BISHOP, ROOK, QUEEN};
	public static final int[] PIECE_VALUES_OE = {0, oe(PAWN, PAWN), oe(KNIGHT, KNIGHT), oe(BISHOP, BISHOP), oe(ROOK, ROOK), oe(QUEEN, QUEEN)};
	public static final int BISHOP_PAIR = oe(50, 50); // Bonus by having two bishops in different colors

	public static final int GAME_PHASE_MIDGAME = 1000;
	public static final int GAME_PHASE_ENDGAME = 0;
	public static final int NON_PAWN_MATERIAL_ENDGAME_MIN = QUEEN + ROOK;
	public static final int NON_PAWN_MATERIAL_MIDGAME_MAX = 3 * KNIGHT + 3 * BISHOP + 4 * ROOK + 2 * QUEEN;

	public BitboardAttacks bbAttacks;

	public Evaluator() {
		bbAttacks = BitboardAttacks.getInstance();
	}

	/**
	 * Board evaluator
	 */
	public abstract int evaluate(Board board, AttacksInfo attacksInfo);

	/**
	 * Merges two short Opening - Ending values in one int
	 */
	public static int oe(int o, int e) {
		return (o << 16) + e;
	}

	/**
	 * Get the "Opening" part
	 */
	public static int o(int oe) {
		return (oe + 0x8000) >> 16;
	}

	/**
	 * Get the "Endgame" part
	 */
	public static int e(int oe) {
		return (short) (oe & 0xffff);
	}

	/**
	 * Shift right each part by factor positions
	 */
	public static int oeShr(int factor, int oeValue) {
		return oe(o(oeValue) >> factor, e(oeValue) >> factor);
	}

	String formatOE(int value) {
		return StringUtils.padLeft(String.valueOf(o(value)), 8) + " " + StringUtils.padLeft(String.valueOf(e(value)), 8);
	}

}

File: csharp/Carballo/Com/Alonsoruibal/Chess/Evaluation/Evaluator.cs
/*
 * Code converted with Sharpen
 * 
 */using Com.Alonsoruibal.Chess;
using Com.Alonsoruibal.Chess.Bitboard;
using Sharpen;

namespace Com.Alonsoruibal.Chess.Evaluation
{
	public abstract class Evaluator
	{
		public const int VICTORY = short.MaxValue - 1;

		public BitboardAttacks bbAttacks;

		public Evaluator()
		{
			//	private final static Random random = new Random(System.currentTimeMillis());
			bbAttacks = BitboardAttacks.GetInstance();
		}

		/// <summary>Board evaluator</summary>
		public abstract int EvaluateBoard(Board board, int alpha, int beta);

		public static int Oe(int opening, int endgame)
		{
			return (((short)(opening)) << 16) + (short)(endgame);
		}

		public static int O(int oe)
		{
			return oe >> 16;
		}

		public static int E(int oe)
		{
			return (short)(oe & unchecked((int)(0xffff)));
		}
	}
}
